xpra/server/auth/keycloak_auth.py

```diff
@@ -0,0 +1,125 @@
+# This file is part of Xpra.
+# Copyright (C) 2016-2021 Antoine Martin <antoine@xpra.org>
+# Copyright (C) 2022 Nathalie Casati <nat@yuka.ch>
+# Xpra is released under the terms of the GNU GPL v2, or, at your option, any
+# later version. See the file COPYING for details.
+
+import os
+import json
+
+from xpra.util import typedict
+from xpra.server.auth.sys_auth_base import SysAuthenticator, log
+from keycloak import KeycloakOpenID
+from oauthlib.oauth2 import WebApplicationClient
+
+KEYCLOAK_SERVER_URL = os.environ.get("XPRA_KEYCLOAK_SERVER_URL", "http://localhost:8080/auth/")
+KEYCLOAK_REALM_NAME = os.environ.get("XPRA_KEYCLOAK_REALM_NAME", "example_realm")
+KEYCLOAK_CLIENT_ID  = os.environ.get("XPRA_KEYCLOAK_CLIENT_ID", "example_client")
+KEYCLOAK_CLIENT_SECRET_KEY = os.environ.get("XPRA_KEYCLOAK_CLIENT_SECRET_KEY", "secret")
+KEYCLOAK_REDIRECT_URI = os.environ.get("XPRA_KEYCLOAK_REDIRECT_URI", "http://localhost/login/")
+KEYCLOAK_SCOPE = os.environ.get("XPRA_KEYCLOAK_SCOPE", "openid")
+KEYCLOAK_GRANT_TYPE = os.environ.get("XPRA_KEYCLOAK_GRANT_TYPE", "authorization_code")
+
+
+class Authenticator(SysAuthenticator):
+
+    def __init__(self, **kwargs):
+        self.server_url = kwargs.pop("server_url", KEYCLOAK_SERVER_URL)
+        self.realm_name = kwargs.pop("realm_name", KEYCLOAK_REALM_NAME)
+        self.client_id = kwargs.pop("client_id", KEYCLOAK_CLIENT_ID)
+        self.client_secret_key = kwargs.pop("client_secret_key", KEYCLOAK_CLIENT_SECRET_KEY)
+        self.redirect_uri = kwargs.pop("redirect_uri", KEYCLOAK_REDIRECT_URI)
+        self.scope = kwargs.pop("scope", KEYCLOAK_SCOPE)
+        self.grant_type = kwargs.pop("grant_type", KEYCLOAK_GRANT_TYPE)
+        kwargs["prompt"] = kwargs.pop("prompt", "keycloak")
+
+        if KEYCLOAK_GRANT_TYPE == "authorization_code":
+          super().__init__(**kwargs)
+          log("keycloak auth: server_url=%s, client_id=%s, realm_name=%s, redirect_uri=%s, scope=%s, grant_type=%s",
+              self.server_url, self.client_id, self.realm_name, self.redirect_uri, self.scope, self.grant_type)
+
+          # Get authorization code
+          client = WebApplicationClient(KEYCLOAK_CLIENT_ID)
+          authorization_url = KEYCLOAK_SERVER_URL + 'realms/' + KEYCLOAK_REALM_NAME + '/protocol/openid-connect/auth'
+          self.salt = client.prepare_request_uri(
+                      authorization_url,
+                      redirect_uri = KEYCLOAK_REDIRECT_URI,
+                      scope = [KEYCLOAK_SCOPE],
+          )
+        else:
+          raise(NotImplementedError("Warning: only grant type \"authorization_code\" is currently supported."))
+
+    def __repr__(self):
+        return "keycloak"
+
+    def get_challenge(self, digests):
+        assert not self.challenge_sent
+        if "keycloak" not in digests:
+            log.error("Error: client does not support keycloak authentication")
+            return None
+        self.challenge_sent = True
+        return self.salt, "keycloak"
```

Comment:
Shouldn't this be `password` since there is no `keycloak` challenge handler?

Reply:
No, this is correct. We are using this in the client. There is no password.

Reply:
Just thinking out loud here now that I have figured out what I meant here.

---

The `token` is the "password" in this context (client auth modules) and it is sent unencrypted (see the change to `xpra/net/digest.py`). That's because the token is not meant to be re-usable from another client - how that's enforced is the responsibility of `keycloak`.

What bothers me here is not new to this authentication module so this is not a showstopper: the client will be prompting the user for a "keycloak" (and I think it should say `keycloak token` instead) and then sending it unencrypted without telling the user that this is the case, which could end up leaking a real password the user types in accidentally. (yes, users will do stupid things - I do too)

The `kerberos` and `gss` client handlers acquire the token themselves:
https://github.com/Xpra-org/xpra/blob/137a222006ce28cd18d8def7884e9d6aeb89b38b/xpra/client/auth/kerberos_handler.py#L76
https://github.com/Xpra-org/xpra/blob/137a222006ce28cd18d8def7884e9d6aeb89b38b/xpra/client/auth/gss_handler.py#L54
And I was hoping that a `keycloak` client authentication module could do the same thing.
But the problem with all those modules is that when they fail, we end up also calling the `prompt` client auth handler as fallback..

The `prompt` handler even has an ugly default prompt specific for `kerberos` and `gss`:
https://github.com/Xpra-org/xpra/blob/137a222006ce28cd18d8def7884e9d6aeb89b38b/xpra/client/auth/prompt_handler.py#L24-L25

I am just thinking that it should be more obvious to the user that what they type in the prompt handler can be intercepted in this case.
(and obviously skip this warning if the connection to the server is actually encrypted with `ssl`, `wss`, `ssh` or `aes`)